### DISCUSSION TITLE
How to make mockery detect typos in foo.On("MyMethox")?

### DISCUSSION THREAD

**Discussion by guettli** [vektra/mockery#765] [Q&A] [answered]
In my setup v2.35.3 mockery does not complain, if I have a type like `foo.On("MyMethox")`.

Or if I don't provide an argument to a method which requires an argument.

If the method of the interface requires a string, and I just use foo.On("MethodWhichWantsAString"), then this gets silently ignored.

This can waste precious time.

Is there a way to make mockery complain typos and missing arguments? about that?

**Comment by LandonTClipp** [comment, marked as answer]
Yes, usage of the `.EXPECT()` builder pattern. This returns a struct with methods that are (mostly) identical to your interface. So one of its methods would be, in your example, `.MethodWhichWantsAString("foobar")`. This will automatically call out the appropriate expectation in testify. This method is a bit more type safe as if you can't misspell the method name, and if you miss an argument, you'll get compiler errors.

It's not 100% fool proof because the argument types in the method are `interface{}` because we need a way to specify `mock.Anything`. A way to get around that is something like `mock.EXPECT().MethodWhichWantsAString(mock.Anything).RunAndReturn(func(s string){ return "foo" })`. This is the most type-safe way to specify expectations.

**Reply by guettli** [reply]
Thank you for the hint. I the code base I work only only `.On()` gets used, so I was not aware of `.EXPECT()`.

I should have read the [docs](https://vektra.github.io/mockery/latest/examples/):

> We recommend always interacting with the assertions through .EXPECT() as mockery auto-generates methods that call out to Mock.On() themselves, providing you with some amount of compile-time safety. Consider
